File: internal/ast/ast.go
package ast

import (
	"fmt"
	"strings"
	"yar/internal/token"
)

type Node interface {
	Pos() token.Position
}

type Declaration interface {
	Node
	declNode()
}

type Statement interface {
	Node
	stmtNode()
}

type Expression interface {
	Node
	exprNode()
}

type Program struct {
	PackagePos  token.Position
	PackageName string
	Imports     []ImportDecl
	Structs     []*StructDecl
	Interfaces  []*InterfaceDecl
	Enums       []*EnumDecl
	Functions   []*FunctionDecl
}

func (p *Program) Pos() token.Position {
	return p.PackagePos
}

type TypeRefKind int

const (
	NamedTypeRef TypeRefKind = iota
	PointerTypeRef
	ArrayTypeRef
	SliceTypeRef
	MapTypeRef
	FunctionTypeRef
)

type TypeRef struct {
	Kind      TypeRefKind
	Name      string
	TypeArgs  []TypeRef
	ArrayLen  int
	Elem      *TypeRef
	Key       *TypeRef
	Value     *TypeRef
	Params    []TypeRef
	Return    *TypeRef
	Errorable bool
	Pos       token.Position
}

func (r TypeRef) String() string {
	switch r.Kind {
	case PointerTypeRef:
		if r.Elem == nil {
			return "*"
		}
		return "*" + r.Elem.String()
	case ArrayTypeRef:
		if r.Elem == nil {
			return fmt.Sprintf("[%d]", r.ArrayLen)
		}
		return fmt.Sprintf("[%d]%s", r.ArrayLen, r.Elem.String())
	case SliceTypeRef:
		if r.Elem == nil {
			return "[]"
		}
		return "[]" + r.Elem.String()
	case MapTypeRef:
		var key, value string
		if r.Key != nil {
			key = r.Key.String()
		}
		if r.Value != nil {
			value = r.Value.String()
		}
		return "map[" + key + "]" + value
	case FunctionTypeRef:
		params := make([]string, 0, len(r.Params))
		for _, param := range r.Params {
			params = append(params, param.String())
		}
		ret := ""
		if r.Return != nil {
			ret = r.Return.String()
		}
		if r.Errorable {
			ret = "!" + ret
		}
		return "fn(" + strings.Join(params, ", ") + ") " + ret
	default:
		if len(r.TypeArgs) == 0 {
			return r.Name
		}
		parts := make([]string, 0, len(r.TypeArgs))
		for _, arg := range r.TypeArgs {
			parts = append(parts, arg.String())
		}
		return r.Name + "[" + strings.Join(parts, ", ") + "]"
	}
}

type ImportDecl struct {
	ImportPos token.Position
	Path      string
	PathPos   token.Position
}

type TypeParam struct {
	Name string
	Pos  token.Position
}

type StructDecl struct {
	StructPos  token.Position
	Exported   bool
	Name       string
	NamePos    token.Position
	TypeParams []TypeParam
	Fields     []StructField
}

func (d *StructDecl) Pos() token.Position {
	return d.StructPos
}

func (*StructDecl) declNode() {}

type StructField struct {
	Name    string
	NamePos token.Position
	Type    TypeRef
}

type InterfaceDecl struct {
	InterfacePos token.Position
	Exported     bool
	Name         string
	NamePos      token.Position
	Methods      []InterfaceMethodDecl
}

func (d *InterfaceDecl) Pos() token.Position {
	return d.InterfacePos
}

func (*InterfaceDecl) declNode() {}

type InterfaceMethodDecl struct {
	Name         string
	NamePos      token.Position
	Params       []Param
	Return       TypeRef
	ReturnIsBang bool
}

type EnumDecl struct {
	EnumPos  token.Position
	Exported bool
	Name     string
	NamePos  token.Position
	Cases    []EnumCaseDecl
}

func (d *EnumDecl) Pos() token.Position {
	return d.EnumPos
}

func (*EnumDecl) declNode() {}

type EnumCaseDecl struct {
	Name    string
	NamePos token.Position
	Fields  []StructField
}

type FunctionDecl struct {
	Exported     bool
	Name         string
	NamePos      token.Position
	TypeParams   []TypeParam
	Receiver     *ReceiverDecl
	Params       []Param
	Return       TypeRef
	ReturnIsBang bool
	Body         *BlockStmt
}

func (f *FunctionDecl) Pos() token.Position {
	return f.NamePos
}

func (*FunctionDecl) declNode() {}

type Param struct {
	Name    string
	NamePos token.Position
	Type    TypeRef
}

type ReceiverDecl struct {
	Name    string
	NamePos token.Position
	Type    TypeRef
}

type BlockStmt struct {
	LBrace token.Position
	Stmts  []Statement
}

func (b *BlockStmt) Pos() token.Position {
	return b.LBrace
}

func (*BlockStmt) stmtNode() {}

type LetStmt struct {
	LetPos  token.Position
	Name    string
	NamePos token.Position
	Value   Expression
}

func (s *LetStmt) Pos() token.Position {
	return s.LetPos
}

func (*LetStmt) stmtNode() {}

type VarStmt struct {
	VarPos  token.Position
	Name    string
	NamePos token.Position
	Type    TypeRef
	Value   Expression
}

func (s *VarStmt) Pos() token.Position {
	return s.VarPos
}

func (*VarStmt) stmtNode() {}

type AssignStmt struct {
	Target Expression
	Value  Expression
}

func (s *AssignStmt) Pos() token.Position {
	return s.Target.Pos()
}

func (*AssignStmt) stmtNode() {}

type IfStmt struct {
	IfPos token.Position
	Cond  Expression
	Then  *BlockStmt
	Else  Statement
}

func (s *IfStmt) Pos() token.Position {
	return s.IfPos
}

func (*IfStmt) stmtNode() {}

type ForStmt struct {
	ForPos token.Position
	Init   Statement
	Cond   Expression
	Post   Statement
	Body   *BlockStmt
}

func (s *ForStmt) Pos() token.Position {
	return s.ForPos
}

func (*ForStmt) stmtNode() {}

type BreakStmt struct {
	BreakPos token.Position
}

func (s *BreakStmt) Pos() token.Position {
	return s.BreakPos
}

func (*BreakStmt) stmtNode() {}

type ContinueStmt struct {
	ContinuePos token.Position
}

func (s *ContinueStmt) Pos() token.Position {
	return s.ContinuePos
}

func (*ContinueStmt) stmtNode() {}

type ReturnStmt struct {
	ReturnPos token.Position
	Value     Expression
}

func (s *ReturnStmt) Pos() token.Position {
	return s.ReturnPos
}

func (*ReturnStmt) stmtNode() {}

type MatchStmt struct {
	MatchPos token.Position
	Value    Expression
	Arms     []MatchArm
}

func (s *MatchStmt) Pos() token.Position {
	return s.MatchPos
}

func (*MatchStmt) stmtNode() {}

type MatchArm struct {
	CasePos     token.Position
	EnumType    TypeRef
	CaseName    string
	CaseNamePos token.Position
	BindName    string
	BindNamePos token.Position
	BindIgnore  bool
	Body        *BlockStmt
}

type ExprStmt struct {
	Expr Expression
}

func (s *ExprStmt) Pos() token.Position {
	return s.Expr.Pos()
}

func (*ExprStmt) stmtNode() {}

type IdentExpr struct {
	Name    string
	NamePos token.Position
}

func (e *IdentExpr) Pos() token.Position {
	return e.NamePos
}

func (*IdentExpr) exprNode() {}

type IntLiteral struct {
	Value  int64
	LitPos token.Position
}

func (e *IntLiteral) Pos() token.Position {
	return e.LitPos
}

func (*IntLiteral) exprNode() {}

type StringLiteral struct {
	Value  string
	LitPos token.Position
}

func (e *StringLiteral) Pos() token.Position {
	return e.LitPos
}

func (*StringLiteral) exprNode() {}

type BoolLiteral struct {
	Value  bool
	LitPos token.Position
}

func (e *BoolLiteral) Pos() token.Position {
	return e.LitPos
}

func (*BoolLiteral) exprNode() {}

type NilLiteral struct {
	LitPos token.Position
}

func (e *NilLiteral) Pos() token.Position {
	return e.LitPos
}

func (*NilLiteral) exprNode() {}

type ErrorLiteral struct {
	Name   string
	ErrPos token.Position
}

func (e *ErrorLiteral) Pos() token.Position {
	return e.ErrPos
}

func (*ErrorLiteral) exprNode() {}

type TypeApplicationExpr struct {
	Inner       Expression
	LBracketPos token.Position
	TypeArgs    []TypeRef
}

func (e *TypeApplicationExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*TypeApplicationExpr) exprNode() {}

type CallExpr struct {
	Callee Expression
	Args   []Expression
}

func (e *CallExpr) Pos() token.Position {
	return e.Callee.Pos()
}

func (*CallExpr) exprNode() {}

type FunctionLiteralExpr struct {
	FnPos        token.Position
	Params       []Param
	Return       TypeRef
	ReturnIsBang bool
	Body         *BlockStmt
}

func (e *FunctionLiteralExpr) Pos() token.Position {
	return e.FnPos
}

func (*FunctionLiteralExpr) exprNode() {}

type UnaryExpr struct {
	Operator token.Kind
	OpPos    token.Position
	Inner    Expression
}

func (e *UnaryExpr) Pos() token.Position {
	return e.OpPos
}

func (*UnaryExpr) exprNode() {}

type BinaryExpr struct {
	Left     Expression
	Operator token.Kind
	OpPos    token.Position
	Right    Expression
}

func (e *BinaryExpr) Pos() token.Position {
	return e.Left.Pos()
}

func (*BinaryExpr) exprNode() {}

type GroupExpr struct {
	Inner Expression
}

func (e *GroupExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*GroupExpr) exprNode() {}

type SelectorExpr struct {
	Inner   Expression
	DotPos  token.Position
	Name    string
	NamePos token.Position
}

func (e *SelectorExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*SelectorExpr) exprNode() {}

type IndexExpr struct {
	Inner       Expression
	LBracketPos token.Position
	Index       Expression
}

func (e *IndexExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*IndexExpr) exprNode() {}

type SliceExpr struct {
	Inner       Expression
	LBracketPos token.Position
	Start       Expression
	ColonPos    token.Position
	End         Expression
}

func (e *SliceExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*SliceExpr) exprNode() {}

type StructLiteralExpr struct {
	Type   TypeRef
	LBrace token.Position
	Fields []StructLiteralField
}

func (e *StructLiteralExpr) Pos() token.Position {
	return e.Type.Pos
}

func (*StructLiteralExpr) exprNode() {}

type StructLiteralField struct {
	Name    string
	NamePos token.Position
	Value   Expression
}

type ArrayLiteralExpr struct {
	Type     TypeRef
	LBrace   token.Position
	Elements []Expression
}

func (e *ArrayLiteralExpr) Pos() token.Position {
	return e.Type.Pos
}

func (*ArrayLiteralExpr) exprNode() {}

type SliceLiteralExpr struct {
	Type     TypeRef
	LBrace   token.Position
	Elements []Expression
}

func (e *SliceLiteralExpr) Pos() token.Position {
	return e.Type.Pos
}

func (*SliceLiteralExpr) exprNode() {}

type MapLiteralExpr struct {
	Type   TypeRef
	LBrace token.Position
	Pairs  []MapLiteralPair
}

func (e *MapLiteralExpr) Pos() token.Position {
	return e.Type.Pos
}

func (*MapLiteralExpr) exprNode() {}

type MapLiteralPair struct {
	Key      Expression
	KeyPos   token.Position
	Value    Expression
	ValuePos token.Position
}

type PropagateExpr struct {
	Inner       Expression
	QuestionPos token.Position
}

func (e *PropagateExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*PropagateExpr) exprNode() {}

type HandleExpr struct {
	Inner   Expression
	OrPos   token.Position
	ErrName string
	ErrPos  token.Position
	Handler *BlockStmt
}

func (e *HandleExpr) Pos() token.Position {
	return e.Inner.Pos()
}

func (*HandleExpr) exprNode() {}

File: crates/yar-compiler/src/ast.rs
use std::{collections::BTreeMap, fmt};

use crate::token::{Kind, Position};

#[derive(Clone, Debug, Default, PartialEq)]
pub struct Program {
    pub package_pos: Position,
    pub package_name: String,
    pub imports: Vec<ImportDecl>,
    pub structs: Vec<StructDecl>,
    pub interfaces: Vec<InterfaceDecl>,
    pub enums: Vec<EnumDecl>,
    pub functions: Vec<FunctionDecl>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct PackageImport {
    pub name: String,
    pub path: String,
    pub target: PackageId,
    pub decl: ImportDecl,
}

#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub enum SourceId {
    #[default]
    Entry,
    Path {
        manifest_path: String,
    },
    Git {
        git: String,
        commit: String,
    },
    Stdlib,
}

/// Stable package identity inside one compilation graph.
///
/// Import spellings and dependency aliases are bindings owned by a source;
/// they are deliberately not part of the package identity.
#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct PackageId {
    pub source: SourceId,
    pub subpath: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct Package {
    pub id: PackageId,
    pub name: String,
    pub stdlib: bool,
    pub files: Vec<Program>,
    pub imports: Vec<PackageImport>,
    pub structs: Vec<StructDecl>,
    pub interfaces: Vec<InterfaceDecl>,
    pub enums: Vec<EnumDecl>,
    pub functions: Vec<FunctionDecl>,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PackageGraph {
    pub entry: PackageId,
    pub packages: BTreeMap<PackageId, Package>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum TypeRefKind {
    Named,
    Errorable,
    Pointer,
    Array,
    Slice,
    Map,
    Chan,
    Function,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TypeRef {
    pub kind: TypeRefKind,
    pub name: String,
    pub type_args: Vec<TypeRef>,
    pub array_len: i64,
    pub elem: Option<Box<TypeRef>>,
    pub key: Option<Box<TypeRef>>,
    pub value: Option<Box<TypeRef>>,
    pub params: Vec<TypeRef>,
    pub return_type: Option<Box<TypeRef>>,
    pub errorable: bool,
    pub pos: Position,
}

impl TypeRef {
    pub fn named(name: impl Into<String>, pos: Position) -> Self {
        Self {
            kind: TypeRefKind::Named,
            name: name.into(),
            type_args: Vec::new(),
            array_len: 0,
            elem: None,
            key: None,
            value: None,
            params: Vec::new(),
            return_type: None,
            errorable: false,
            pos,
        }
    }
}

impl fmt::Display for TypeRef {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.kind {
            TypeRefKind::Errorable => write_optional_prefix(f, "!", self.elem.as_deref()),
            TypeRefKind::Pointer => write_optional_prefix(f, "*", self.elem.as_deref()),
            TypeRefKind::Array => {
                write!(f, "[{}]", self.array_len)?;
                if let Some(elem) = self.elem.as_deref() {
                    write!(f, "{elem}")?;
                }
                Ok(())
            }
            TypeRefKind::Slice => write_optional_prefix(f, "[]", self.elem.as_deref()),
            TypeRefKind::Map => {
                write!(f, "map[")?;
                if let Some(key) = self.key.as_deref() {
                    write!(f, "{key}")?;
                }
                write!(f, "]")?;
                if let Some(value) = self.value.as_deref() {
                    write!(f, "{value}")?;
                }
                Ok(())
            }
            TypeRefKind::Chan => {
                write!(f, "chan[")?;
                if let Some(elem) = self.elem.as_deref() {
                    write!(f, "{elem}")?;
                }
                write!(f, "]")
            }
            TypeRefKind::Function => {
                let params = self
                    .params
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<_>>()
                    .join(", ");
                write!(f, "fn({params}) ")?;
                if self.errorable {
                    write!(f, "!")?;
                }
                if let Some(return_type) = self.return_type.as_deref() {
                    write!(f, "{return_type}")?;
                }
                Ok(())
            }
            TypeRefKind::Named => {
                if self.type_args.is_empty() {
                    return f.write_str(&self.name);
                }
                let args = self
                    .type_args
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<_>>()
                    .join(", ");
                write!(f, "{}[{args}]", self.name)
            }
        }
    }
}

fn write_optional_prefix(
    f: &mut fmt::Formatter<'_>,
    prefix: &str,
    elem: Option<&TypeRef>,
) -> fmt::Result {
    f.write_str(prefix)?;
    if let Some(elem) = elem {
        write!(f, "{elem}")?;
    }
    Ok(())
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ImportDecl {
    pub import_pos: Position,
    pub path: String,
    pub path_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TypeParam {
    pub name: String,
    pub pos: Position,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructDecl {
    pub struct_pos: Position,
    pub exported: bool,
    pub resource: bool,
    pub name: String,
    pub name_pos: Position,
    pub type_params: Vec<TypeParam>,
    pub fields: Vec<StructField>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructField {
    pub name: String,
    pub name_pos: Position,
    pub type_ref: TypeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct InterfaceDecl {
    pub interface_pos: Position,
    pub exported: bool,
    pub name: String,
    pub name_pos: Position,
    pub methods: Vec<InterfaceMethodDecl>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct InterfaceMethodDecl {
    pub name: String,
    pub name_pos: Position,
    pub params: Vec<Param>,
    pub return_type: TypeRef,
    pub return_is_bang: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EnumDecl {
    pub enum_pos: Position,
    pub exported: bool,
    pub name: String,
    pub name_pos: Position,
    pub cases: Vec<EnumCaseDecl>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EnumCaseDecl {
    pub name: String,
    pub name_pos: Position,
    pub fields: Vec<StructField>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionDecl {
    pub exported: bool,
    pub host_intrinsic: bool,
    pub name: String,
    pub name_pos: Position,
    pub type_params: Vec<TypeParam>,
    pub receiver: Option<ReceiverDecl>,
    pub params: Vec<Param>,
    pub return_type: TypeRef,
    pub return_is_bang: bool,
    pub body: BlockStmt,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Param {
    pub name: String,
    pub name_pos: Position,
    pub type_ref: TypeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReceiverDecl {
    pub name: String,
    pub name_pos: Position,
    pub type_ref: TypeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BlockStmt {
    pub lbrace: Position,
    pub stmts: Vec<Statement>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Statement {
    Block(Box<BlockStmt>),
    Let(Box<LetStmt>),
    Var(Box<VarStmt>),
    Assign(Box<AssignStmt>),
    CompoundAssign(Box<CompoundAssignStmt>),
    If(Box<IfStmt>),
    For(Box<ForStmt>),
    Break(BreakStmt),
    Continue(ContinueStmt),
    Return(Box<ReturnStmt>),
    Match(Box<MatchStmt>),
    Expr(Box<ExprStmt>),
    Spawn(Box<SpawnStmt>),
}

#[derive(Clone, Debug, PartialEq)]
pub struct LetStmt {
    pub let_pos: Position,
    pub name: String,
    pub name_pos: Position,
    pub value: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct VarStmt {
    pub var_pos: Position,
    pub name: String,
    pub name_pos: Position,
    pub type_ref: TypeRef,
    pub value: Option<Expression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct AssignStmt {
    pub target: Expression,
    pub value: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct CompoundAssignStmt {
    pub target: Expression,
    pub operator: Kind,
    pub op_pos: Position,
    pub value: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct IfStmt {
    pub if_pos: Position,
    pub cond: Expression,
    pub then_block: BlockStmt,
    pub else_stmt: Option<Statement>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ForStmt {
    pub for_pos: Position,
    pub init: Option<Statement>,
    pub cond: Option<Expression>,
    pub post: Option<Statement>,
    pub body: BlockStmt,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BreakStmt {
    pub break_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ContinueStmt {
    pub continue_pos: Position,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReturnStmt {
    pub return_pos: Position,
    pub value: Option<Expression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct MatchStmt {
    pub match_pos: Position,
    pub value: Expression,
    pub arms: Vec<MatchArm>,
    pub else_body: Option<BlockStmt>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct MatchArm {
    pub case_pos: Position,
    pub enum_type: TypeRef,
    pub case_name: String,
    pub case_name_pos: Position,
    pub bind_name: String,
    pub bind_name_pos: Position,
    pub bind_ignore: bool,
    pub body: BlockStmt,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ExprStmt {
    pub expr: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SpawnStmt {
    pub spawn_pos: Position,
    pub call: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Expression {
    Ident(Box<IdentExpr>),
    Int(IntLiteral),
    Char(CharLiteral),
    String(StringLiteral),
    Bool(BoolLiteral),
    Nil(NilLiteral),
    Error(ErrorLiteral),
    TypeApplication(Box<TypeApplicationExpr>),
    Call(Box<CallExpr>),
    FunctionLiteral(Box<FunctionLiteralExpr>),
    Taskgroup(Box<TaskgroupExpr>),
    Unary(Box<UnaryExpr>),
    Binary(Box<BinaryExpr>),
    Group(Box<GroupExpr>),
    Selector(Box<SelectorExpr>),
    Index(Box<IndexExpr>),
    Slice(Box<SliceExpr>),
    StructLiteral(Box<StructLiteralExpr>),
    ArrayLiteral(Box<ArrayLiteralExpr>),
    SliceLiteral(Box<SliceLiteralExpr>),
    MapLiteral(Box<MapLiteralExpr>),
    Propagate(Box<PropagateExpr>),
    Handle(Box<HandleExpr>),
    Missing(Position),
}

impl Expression {
    pub fn pos(&self) -> Position {
        match self {
            Expression::Ident(expr) => expr.name_pos.clone(),
            Expression::Int(expr) => expr.lit_pos.clone(),
            Expression::Char(expr) => expr.lit_pos.clone(),
            Expression::String(expr) => expr.lit_pos.clone(),
            Expression::Bool(expr) => expr.lit_pos.clone(),
            Expression::Nil(expr) => expr.lit_pos.clone(),
            Expression::Error(expr) => expr.err_pos.clone(),
            Expression::TypeApplication(expr) => expr.inner.pos(),
            Expression::Call(expr) => expr.callee.pos(),
            Expression::FunctionLiteral(expr) => expr.fn_pos.clone(),
            Expression::Taskgroup(expr) => expr.taskgroup_pos.clone(),
            Expression::Unary(expr) => expr.op_pos.clone(),
            Expression::Binary(expr) => expr.left.pos(),
            Expression::Group(expr) => expr.inner.pos(),
            Expression::Selector(expr) => expr.inner.pos(),
            Expression::Index(expr) => expr.inner.pos(),
            Expression::Slice(expr) => expr.inner.pos(),
            Expression::StructLiteral(expr) => expr.type_ref.pos.clone(),
            Expression::ArrayLiteral(expr) => expr.type_ref.pos.clone(),
            Expression::SliceLiteral(expr) => expr.type_ref.pos.clone(),
            Expression::MapLiteral(expr) => expr.type_ref.pos.clone(),
            Expression::Propagate(expr) => expr.inner.pos(),
            Expression::Handle(expr) => expr.inner.pos(),
            Expression::Missing(pos) => pos.clone(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct IdentExpr {
    pub name: String,
    pub name_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct IntLiteral {
    pub value: i64,
    pub lit_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CharLiteral {
    pub value: char,
    pub lit_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct StringLiteral {
    pub value: String,
    pub lit_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BoolLiteral {
    pub value: bool,
    pub lit_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct NilLiteral {
    pub lit_pos: Position,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ErrorLiteral {
    pub name: String,
    pub err_pos: Position,
}

#[derive(Clone, Debug, PartialEq)]
pub struct TypeApplicationExpr {
    pub inner: Expression,
    pub lbracket_pos: Position,
    pub type_args: Vec<TypeRef>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct CallExpr {
    pub callee: Expression,
    pub args: Vec<Expression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionLiteralExpr {
    pub fn_pos: Position,
    pub enclosing_function: String,
    pub params: Vec<Param>,
    pub return_type: TypeRef,
    pub return_is_bang: bool,
    pub body: BlockStmt,
}

#[derive(Clone, Debug, PartialEq)]
pub struct TaskgroupExpr {
    pub taskgroup_pos: Position,
    pub result_type: TypeRef,
    pub body: BlockStmt,
}

#[derive(Clone, Debug, PartialEq)]
pub struct UnaryExpr {
    pub operator: Kind,
    pub op_pos: Position,
    pub inner: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BinaryExpr {
    pub left: Expression,
    pub operator: Kind,
    pub op_pos: Position,
    pub right: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct GroupExpr {
    pub inner: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SelectorExpr {
    pub inner: Expression,
    pub dot_pos: Position,
    pub name: String,
    pub name_pos: Position,
}

#[derive(Clone, Debug, PartialEq)]
pub struct IndexExpr {
    pub inner: Expression,
    pub lbracket_pos: Position,
    pub index: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SliceExpr {
    pub inner: Expression,
    pub lbracket_pos: Position,
    pub start: Option<Expression>,
    pub colon_pos: Position,
    pub end: Option<Expression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructLiteralExpr {
    pub type_ref: TypeRef,
    pub lbrace: Position,
    pub fields: Vec<StructLiteralField>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructLiteralField {
    pub name: String,
    pub name_pos: Position,
    pub value: Expression,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrayLiteralExpr {
    pub type_ref: TypeRef,
    pub lbrace: Position,
    pub elements: Vec<Expression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SliceLiteralExpr {
    pub type_ref: TypeRef,
    pub lbrace: Position,
    pub elements: Vec<Expression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct MapLiteralExpr {
    pub type_ref: TypeRef,
    pub lbrace: Position,
    pub pairs: Vec<MapLiteralPair>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct MapLiteralPair {
    pub key: Expression,
    pub key_pos: Position,
    pub value: Expression,
    pub value_pos: Position,
}

#[derive(Clone, Debug, PartialEq)]
pub struct PropagateExpr {
    pub inner: Expression,
    pub question_pos: Position,
}

#[derive(Clone, Debug, PartialEq)]
pub struct HandleExpr {
    pub inner: Expression,
    pub or_pos: Position,
    pub err_name: String,
    pub err_pos: Position,
    pub handler: BlockStmt,
}
